core/finality-grandpa/src/environment.rs

```diff
@@ -317,7 +324,7 @@ impl<Block: BlockT<Hash=H256>, B, E, N, RA> grandpa::Chain<Block::Hash, NumberFo
 		let limit = self.authority_set.current_limit();
 		debug!(target: "afg", "Finding best chain containing block {:?} with number limit {:?}", block, limit);
 
-		match self.inner.best_containing(block, None) {
+		match self.select_chain.finality_target(block, None) {
```

Comment:
yeah, I guess the main thing is that we need a way to get the best chain containing a given hash. I thought of doing it as a separate operation, where we would use `sc.finality_target(sc.best_containing(included_hash))`

Reply:
I think the API in this PR is workable for GRANDPA, but I think we need to introduce a split if we want something that would work for finality gadgets and instant-finality authorship and probablistic finality authorship. 

With the current API I'm not sure exactly what a `Tendermint` implementation would build on.

Reply:
I think this was mostly because of the same (wrong) API docs. I updated those.

The way I'd see this work is, that if you have a finality gadget, you'd pass the `best_chain` to `finality_target` and it has some way to find the ancestor that you should vote on.

For instant finality systems (like Tendermint or Rhododendron) I don't see why you'd ever need to use `finality_target` (or any finality gadget for that matter) at all. Because `best_chain` will already return a finalised block, thus the default implementation just returns the same hash again. Same for any probabilistic finality system but the other way around - no block is ever "surely finalized" - thus the system would not use the `finality_target` function - or given a hash and a certain risk profile could return some ancestor it is certain to not be reverted, but as there won't be any finality gadget, no one would ask for it...

Reply:
@gnunicorn I think if we would apply e.g. the polkadot chain-selection rule to an instant-finality authorship system, the behavior we'd desire is that
  - we have a best block which is finalized 
  - but we refuse to propose or vote on a new block until we've received the erasure-coded chunks for that chain. This would be accomplished by doing `finality_target(best_chain()) == best_chain()`.